---
title: "dplyr Lab"
author: ""
date: "July 7, 2015"
output: html_document
---

1. Download the samples data on the 1,000 genomes project from their website (ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/working/20130606_sample_info/20130606_sample_info.xlsx).



2. Open the sheet and look at the information listed. Go to the sheet _Final Phase Sequence Data_

-- this is the 4th sheet

3. Read the _Final Phase Sequence Data_ sheet. Only read the data for the low coverage samples.  

```{r}

library(readxl)
# Have to skip one row because 
# there is an extra header
kg = read_excel("1001genomes.xlsx",
                    sheet=4,skip=1)

kg=kg[1:7] #can't use select because multiple columns have the same name


# make column names easier to handle
colnames(kg) = colnames(kg) %>% tolower() %>% gsub(x=.,pattern=" ",replacement="") %>% substr(1,10)
```

4. Calculate total sequence by platform

```{r}

kg %>% group_by(platform) %>%  summarize(sum(totalseque))

```


5. Do the same thing by sequencing center


```{r}

kg %>% group_by(center) %>%  summarize(sum(totalseque))

```


6. Find the subset of samples that passed QC. 


```{r}

kg %>%  filter(passedqc==1)

```

7.Find the subset that passed QC and came from the BCM center



8. Calculate the average aligned coverage for each population on the subset of samples that passed QC that came from the BCM. 

```{r}
kg %>% filter(passedqc==1) %>%
  filter(center=="BCM") %>%
  group_by(population) %>%
  summarize(meancov = mean(alignednon),nsamp=n())
```
